cylc/flow/parsec/fileparse.py

```diff
@@ -243,23 +263,42 @@ def read_and_proc(fpath, template_vars=None, viewcfg=None, asedit=False):
     if do_empy:
         if flines and re.match(r'^#![Ee]m[Pp]y\s*', flines[0]):
             LOG.debug('Processing with EmPy')
+            tvars = copy(template_vars)
+            if rose_vars['empy:suite.rc'] is not None:
+                for key, value in rose_vars['empy:suite.rc'].items():
+                    tvars[key] = literal_eval(value)
             try:
                 from cylc.flow.parsec.empysupport import empyprocess
             except (ImportError, ModuleNotFoundError):
                 raise ParsecError('EmPy Python package must be installed '
                                   'to process file: ' + fpath)
-            flines = empyprocess(flines, fdir, template_vars)
+            flines = empyprocess(flines, fdir, tvars)
 
     # process with Jinja2
     if do_jinja2:
         if flines and re.match(r'^#![jJ]inja2\s*', flines[0]):
             LOG.debug('Processing with Jinja2')
+            tvars = copy(template_vars)
+            if rose_vars['jinja2:suite.rc'] is not None:
+                for key, value in rose_vars['jinja2:suite.rc'].items():
+                    tvars[key] = literal_eval(value)
```

Comment:
@wxtim, sorry to be a pest, but is `literal_eval` good enough here? There are two ways in which this can break existing suites:
1. Jinja2 code is no longer valid as a value, so things like `range` can't be used anymore. This is probably a very niche use case, so it might be OK to drop it.
2. Environment variables in this section are no longer processed here. This is a much bigger problem, because it's common to have things like `SOME_VAR="\$HOME/some/path"` in this section that are meant to be expanded on the job rather than suite host. Rose will write this as `{% set SOME_VAR="$HOME/some/path" %}`, unlike the above implementation, which will keep the `\$` escape.

An alternative could be to refactor `jinja2:suite.rc` processor in Rose into two methods:
1. Generate `suite.rc` Jinja2 include lines.
2. Write the above lines into `suite.rc`.

This way you could use the first method here directly to inject the Jinja2 lines the way `rose suite-run` did, so the behaviour should be identical. It would also keep the implementation largely where it belongs, i.e. in Rose rather than Cylc, and EmPy would be supported automatically in this way too.

Reply:
[initial response: Will edit comment to add fuller response after reading more carefully] @TomekTrzeciak  - you're not being a pest - I'd rather have this now, from you than later, from some poor sap whose suite has broken in an unexpected way.

Reply:
1. I think this is a positive change.
2. Good spot, if this is the only issue we should be able to work around it easy enough with a bit of string manipulation.

We could really do with ditching the `suite.rc` mangling that Rose2019 does so would avoid writing vars directly into the `suite.rc` at all costs (e.g. you cannot override these vars from the Cylc CLI or any future Cylc files).

Reply:
> The #!Jinja2 shebang is no longer automatically added. 

This sounds kinda fail-safe - if you want jinja2 you _really_ have to want it.

> Rose has some special env var expansion logic built-in to do this, we should be able to import it and bring back this logic easily.

Another PR for another day? I feel mission creep.

Reply:
Can do though it should only be a couple of lines as this was in the original Rose code.

Reply:
> ```shell
> grep -rn --color  --include="rose-suite*.conf" range\(.*\) \
>     /path/to/*manyusers*/conventional_place_to_store_suites/rose-suite*.conf
> ```
> 
> returns one user, a @TomekTrzeciak.... (Oh, and me, now I have a toy suite I made to understand this.

I did find more instances than this, but it is rather uncommon (and as @oliver-sanders points out, might be a good riddance anyway). Mind you, `range` is just an example here, it could be any valid Jinja2 expression like `some_dict.values() | list`.

Reply:
Yes I got the search term of my original grep wrong.

Reply:
> > Rose has some special env var expansion logic built-in to do this, we should be able to import it and bring back this logic easily.
> 
> Another PR for another day? I feel mission creep.

It's just [one import away](https://github.com/metomi/rose/blob/ea60ad9/metomi/rose/config_processors/jinja2.py#L26):

```python
from metomi.rose.env import env_var_process
```

Reply:
Ok, here is my test suite (for MO people /home/h06/osanders/roses/jinja2-test) which I think covers the Rose2019 behaviours that we would want to preserve:

@TomekTrzeciak note 3.3. the escaped env var seems to come through to Jinja2 unescaped though I may have misread your comment above:

*rose-suite.conf*
```ini
[env]
FOO=a
 
[jinja2:suite.rc]
BAR="${FOO}b"
LOCAL_ENV="$XYZ"
ESCAPED_ENV="\$HOME"
INT=42                                                                                                       
BOOL=True
LIST=["a", 1, True]
```

*suite.rc*
```jinja2
# Test with rose `XYZ=xyz suite-run --new -l` (which invokes cylc validate --strict)                         
# Rose env vars are logged to the console
 
# 1. The Jinja2 shebang should be added automatically
{{ assert(True, "failed 1.1") }}
 
##
{% from "subprocess" import Popen %}
{% from "subprocess" import PIPE %}
{# {% from "cylc.flow" import LOG %} #}  # cylc8
{% from "cylc" import LOG %}
 
# 2. ROSE_SUITE_VARIABLES should contain a mapping of all vars
{{ assert(ROSE_SUITE_VARIABLES is defined, "failed 2.1") }}
{{ assert(ROSE_SUITE_VARIABLES["INT"] == 42, "failed 2.2") }}
{% do LOG.info(ROSE_SUITE_VARIABLES) %}
 
# 3. Env vars
# 3.1 - Local env vars should be templated into rose-suite.conf[env]
{{ assert(LOCAL_ENV == "xyz", "failed 3.1") }}
# 3.2 - rose-suite.conf[env] vars should be templated into rose-suite.conf[jinja2:suite.rc]
{{ assert(BAR == "ab", "failed 3.2") }}
# 3.3 - Escaped env vars should not be templated
{{ assert(ESCAPED_ENV == "$HOME", "failed 3.3") }}
 
# 4. Literals should be evaluated
{{ assert(INT == 42, "failed 4.1") }}
{{ assert(BOOL == True, "failed 4.2") }}
{{ assert(LIST == ["a", 1, True], "failed 4.3") }}
 
# 5. ROSE_VERSION should be provided
{{ assert(ROSE_VERSION is defined, "failed 5.1") }}
{% set cli_rose_version = Popen(["rose", "version"], stdout=PIPE).communicate()[0] %}
{% set cli_rose_version = cli_rose_version.replace('Rose ', '') %}  # strip the "Rose " prefix
{% set cli_rose_version = cli_rose_version.split(' ')[0] %}  # strip the (location) suffix
{% do LOG.info("$ rose version: " + cli_rose_version) %}
{{ assert(ROSE_VERSION == cli_rose_version, "failed 5.2") }}
 
# 6. CYLC_VERSION should be provided
{{ assert(CYLC_VERSION is defined, "failed 6.1") }}
{% set cli_cylc_version = Popen(["cylc", "version"], stdout=PIPE).communicate()[0] %}
{% set cli_cylc_version = cli_cylc_version.strip() %}  # remove trailing newline
{% do LOG.info("$ cylc version: " + cli_cylc_version) %}
{{ assert(CYLC_VERSION == cli_cylc_version, "failed 6.2") }}
 
# 7. ROSE_ORIG_HOST should be provided (hard to test it is set correctly here)
{{ assert(ROSE_ORIG_HOST is defined, "failed 7.1") }}
 
# 8. ROSE_SITE should be provided
{{ assert(ROSE_SITE is defined, "failed 8.1") }}
 
 
# chuck in some Cylc stuff to make validate happy
[scheduling]
    initial cycle point = now 
    [[dependencies]]
        [[[R1]]]
            graph = """ 
                x   
            """ 
 
[runtime]
    [[x]]
```

Reply:
> @TomekTrzeciak note 3.3. the escaped env var seems to come through to Jinja2 unescaped though I may have misread your comment above:

Do you mean on Cylc7+Rose2019 or Cylc8+rose-plugin from this PR? I haven't actually tested on this PR (setting up Cylc8 for dev work is too complicated these days), just relying on prior experience with these bits of code. E.g. 3.3 comes through escaped with my own `jinja2render` tool that uses `rose config` to retrieve the vars (should be equivalent to what happens in this PR).

Reply:
The above test suite was written against Rose2019, Cylc7, it does not yet pass against this PR.

Reply:
Thankyou, @oliver-sanders  for writing my functional tests for me. 😄

Reply:
Looks good to me.

A nice interface for implementing this might be:

```
install_plugin() -> {
    'template variables': {},
    'template engine': '',
    ...
}
```

To avoid Cylc having to know about Rose configurations.